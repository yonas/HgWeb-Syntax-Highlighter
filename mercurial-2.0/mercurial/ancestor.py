# ancestor.py - generic DAG ancestor algorithm for mercurial
#
# Copyright 2006 Matt Mackall <mpm@selenic.com>
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.

import heapq

def ancestor(a, b, pfunc):
    """
    Returns the common ancestor of a and b that is furthest from a
    root (as measured by longest path) or None if no ancestor is
    found. If there are multiple common ancestors at the same
    distance, the first one found is returned.

    pfunc must return a list of parent vertices for a given vertex
    """

    if a == b:
        return a

    a, b = sorted([a, b])

    # find depth from root of all ancestors
    # depth is stored as a negative for heapq
    parentcache = {}
    visit = [a, b]
    depth = {}
    while visit:
        vertex = visit[-1]
        pl = pfunc(vertex)
        parentcache[vertex] = pl
        if not pl:
            depth[vertex] = 0
            visit.pop()
        else:
            for p in pl:
                if p == a or p == b: # did we find a or b as a parent?
                    return p # we're done
                if p not in depth:
                    visit.append(p)
            if visit[-1] == vertex:
                # -(maximum distance of parents + 1)
                depth[vertex] = min([depth[p] for p in pl]) - 1
                visit.pop()

    # traverse ancestors in order of decreasing distance from root
    def ancestors(vertex):
        h = [(depth[vertex], vertex)]
        seen = set()
        while h:
            d, n = heapq.heappop(h)
            if n not in seen:
                seen.add(n)
                yield (d, n)
                for p in parentcache[n]:
                    heapq.heappush(h, (depth[p], p))

    def generations(vertex):
        sg, s = None, set()
        for g, v in ancestors(vertex):
            if g != sg:
                if sg:
                    yield sg, s
                sg, s = g, set((v,))
            else:
                s.add(v)
        yield sg, s

    x = generations(a)
    y = generations(b)
    gx = x.next()
    gy = y.next()

    # increment each ancestor list until it is closer to root than
    # the other, or they match
    try:
        while True:
            if gx[0] == gy[0]:
                for v in gx[1]:
                    if v in gy[1]:
                        return v
                gy = y.next()
                gx = x.next()
            elif gx[0] > gy[0]:
                gy = y.next()
            else:
                gx = x.next()
    except StopIteration:
        return None
