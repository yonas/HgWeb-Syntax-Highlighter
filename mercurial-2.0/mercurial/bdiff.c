/*
 bdiff.c - efficient binary diff extension for Mercurial

 Copyright 2005, 2006 Matt Mackall <mpm@selenic.com>

 This software may be used and distributed according to the terms of
 the GNU General Public License, incorporated herein by reference.

 Based roughly on Python difflib
*/

#include <Python.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#if defined __hpux || defined __SUNPRO_C || defined _AIX
#define inline
#endif

#ifdef __linux
#define inline __inline
#endif

#ifdef _WIN32
#ifdef _MSC_VER
#define inline __inline
typedef unsigned long uint32_t;
#else
#include <stdint.h>
#endif
static uint32_t htonl(uint32_t x)
{
	return ((x & 0x000000ffUL) << 24) |
		((x & 0x0000ff00UL) <<  8) |
		((x & 0x00ff0000UL) >>  8) |
		((x & 0xff000000UL) >> 24);
}
#else
#include <sys/types.h>
#if defined __BEOS__ && !defined __HAIKU__
#include <ByteOrder.h>
#else
#include <arpa/inet.h>
#endif
#include <inttypes.h>
#endif

#include "util.h"

struct line {
	int hash, len, n, e;
	const char *l;
};

struct pos {
	int pos, len;
};

struct hunk;
struct hunk {
	int a1, a2, b1, b2;
	struct hunk *next;
};

static int splitlines(const char *a, int len, struct line **lr)
{
	unsigned hash;
	int i;
	const char *p, *b = a;
	const char * const plast = a + len - 1;
	struct line *l;

	/* count the lines */
	i = 1; /* extra line for sentinel */
	for (p = a; p < a + len; p++)
		if (*p == '\n' || p == plast)
			i++;

	*lr = l = (struct line *)malloc(sizeof(struct line) * i);
	if (!l)
		return -1;

	/* build the line array and calculate hashes */
	hash = 0;
	for (p = a; p < a + len; p++) {
		/* Leonid Yuriev's hash */
		hash = (hash * 1664525) + (unsigned char)*p + 1013904223;

		if (*p == '\n' || p == plast) {
			l->hash = hash;
			hash = 0;
			l->len = p - b + 1;
			l->l = b;
			l->n = INT_MAX;
			l++;
			b = p + 1;
		}
	}

	/* set up a sentinel */
	l->hash = 0;
	l->len = 0;
	l->l = a + len;
	return i - 1;
}

static inline int cmp(struct line *a, struct line *b)
{
	return a->hash != b->hash || a->len != b->len || memcmp(a->l, b->l, a->len);
}

static int equatelines(struct line *a, int an, struct line *b, int bn)
{
	int i, j, buckets = 1, t, scale;
	struct pos *h = NULL;

	/* build a hash table of the next highest power of 2 */
	while (buckets < bn + 1)
		buckets *= 2;

	/* try to allocate a large hash table to avoid collisions */
	for (scale = 4; scale; scale /= 2) {
		h = (struct pos *)malloc(scale * buckets * sizeof(struct pos));
		if (h)
			break;
	}

	if (!h)
		return 0;

	buckets = buckets * scale - 1;

	/* clear the hash table */
	for (i = 0; i <= buckets; i++) {
		h[i].pos = INT_MAX;
		h[i].len = 0;
	}

	/* add lines to the hash table chains */
	for (i = bn - 1; i >= 0; i--) {
		/* find the equivalence class */
		for (j = b[i].hash & buckets; h[j].pos != INT_MAX;
		     j = (j + 1) & buckets)
			if (!cmp(b + i, b + h[j].pos))
				break;

		/* add to the head of the equivalence class */
		b[i].n = h[j].pos;
		b[i].e = j;
		h[j].pos = i;
		h[j].len++; /* keep track of popularity */
	}

	/* compute popularity threshold */
	t = (bn >= 31000) ? bn / 1000 : 1000000 / (bn + 1);

	/* match items in a to their equivalence class in b */
	for (i = 0; i < an; i++) {
		/* find the equivalence class */
		for (j = a[i].hash & buckets; h[j].pos != INT_MAX;
		     j = (j + 1) & buckets)
			if (!cmp(a + i, b + h[j].pos))
				break;

		a[i].e = j; /* use equivalence class for quick compare */
		if (h[j].len <= t)
			a[i].n = h[j].pos; /* point to head of match list */
		else
			a[i].n = INT_MAX; /* too popular */
	}

	/* discard hash tables */
	free(h);
	return 1;
}

static int longest_match(struct line *a, struct line *b, struct pos *pos,
			 int a1, int a2, int b1, int b2, int *omi, int *omj)
{
	int mi = a1, mj = b1, mk = 0, mb = 0, i, j, k;

	for (i = a1; i < a2; i++) {
		/* skip things before the current block */
		for (j = a[i].n; j < b1; j = b[j].n)
			;

		/* loop through all lines match a[i] in b */
		for (; j < b2; j = b[j].n) {
			/* does this extend an earlier match? */
			if (i > a1 && j > b1 && pos[j - 1].pos == i - 1)
				k = pos[j - 1].len + 1;
			else
				k = 1;
			pos[j].pos = i;
			pos[j].len = k;

			/* best match so far? */
			if (k > mk) {
				mi = i;
				mj = j;
				mk = k;
			}
		}
	}

	if (mk) {
		mi = mi - mk + 1;
		mj = mj - mk + 1;
	}

	/* expand match to include neighboring popular lines */
	while (mi - mb > a1 && mj - mb > b1 &&
	       a[mi - mb - 1].e == b[mj - mb - 1].e)
		mb++;
	while (mi + mk < a2 && mj + mk < b2 &&
	       a[mi + mk].e == b[mj + mk].e)
		mk++;

	*omi = mi - mb;
	*omj = mj - mb;

	return mk + mb;
}

static struct hunk *recurse(struct line *a, struct line *b, struct pos *pos,
			    int a1, int a2, int b1, int b2, struct hunk *l)
{
	int i, j, k;

	while (1) {
		/* find the longest match in this chunk */
		k = longest_match(a, b, pos, a1, a2, b1, b2, &i, &j);
		if (!k)
			return l;

		/* and recurse on the remaining chunks on either side */
		l = recurse(a, b, pos, a1, i, b1, j, l);
		if (!l)
			return NULL;

		l->next = (struct hunk *)malloc(sizeof(struct hunk));
		if (!l->next)
			return NULL;

		l = l->next;
		l->a1 = i;
		l->a2 = i + k;
		l->b1 = j;
		l->b2 = j + k;
		l->next = NULL;

		/* tail-recursion didn't happen, so do equivalent iteration */
		a1 = i + k;
		b1 = j + k;
	}
}

static int diff(struct line *a, int an, struct line *b, int bn,
		 struct hunk *base)
{
	struct hunk *curr;
	struct pos *pos;
	int t, count = 0;

	/* allocate and fill arrays */
	t = equatelines(a, an, b, bn);
	pos = (struct pos *)calloc(bn ? bn : 1, sizeof(struct pos));

	if (pos && t) {
		/* generate the matching block list */

		curr = recurse(a, b, pos, 0, an, 0, bn, base);
		if (!curr)
			return -1;

		/* sentinel end hunk */
		curr->next = (struct hunk *)malloc(sizeof(struct hunk));
		if (!curr->next)
			return -1;
		curr = curr->next;
		curr->a1 = curr->a2 = an;
		curr->b1 = curr->b2 = bn;
		curr->next = NULL;
	}

	free(pos);

	/* normalize the hunk list, try to push each hunk towards the end */
	for (curr = base->next; curr; curr = curr->next) {
		struct hunk *next = curr->next;
		int shift = 0;

		if (!next)
			break;

		if (curr->a2 == next->a1)
			while (curr->a2 + shift < an && curr->b2 + shift < bn
			       && !cmp(a + curr->a2 + shift,
				       b + curr->b2 + shift))
				shift++;
		else if (curr->b2 == next->b1)
			while (curr->b2 + shift < bn && curr->a2 + shift < an
			       && !cmp(b + curr->b2 + shift,
				       a + curr->a2 + shift))
				shift++;
		if (!shift)
			continue;
		curr->b2 += shift;
		next->b1 += shift;
		curr->a2 += shift;
		next->a1 += shift;
	}

	for (curr = base->next; curr; curr = curr->next)
		count++;
	return count;
}

static void freehunks(struct hunk *l)
{
	struct hunk *n;
	for (; l; l = n) {
		n = l->next;
		free(l);
	}
}

static PyObject *blocks(PyObject *self, PyObject *args)
{
	PyObject *sa, *sb, *rl = NULL, *m;
	struct line *a, *b;
	struct hunk l, *h;
	int an, bn, count, pos = 0;

	if (!PyArg_ParseTuple(args, "SS:bdiff", &sa, &sb))
		return NULL;

	an = splitlines(PyBytes_AsString(sa), PyBytes_Size(sa), &a);
	bn = splitlines(PyBytes_AsString(sb), PyBytes_Size(sb), &b);

	if (!a || !b)
		goto nomem;

	l.next = NULL;
	count = diff(a, an, b, bn, &l);
	if (count < 0)
		goto nomem;

	rl = PyList_New(count);
	if (!rl)
		goto nomem;

	for (h = l.next; h; h = h->next) {
		m = Py_BuildValue("iiii", h->a1, h->a2, h->b1, h->b2);
		PyList_SetItem(rl, pos, m);
		pos++;
	}

nomem:
	free(a);
	free(b);
	freehunks(l.next);
	return rl ? rl : PyErr_NoMemory();
}

static PyObject *bdiff(PyObject *self, PyObject *args)
{
	char *sa, *sb, *rb;
	PyObject *result = NULL;
	struct line *al, *bl;
	struct hunk l, *h;
	uint32_t encode[3];
	int an, bn, len = 0, la, lb, count;

	if (!PyArg_ParseTuple(args, "s#s#:bdiff", &sa, &la, &sb, &lb))
		return NULL;

	an = splitlines(sa, la, &al);
	bn = splitlines(sb, lb, &bl);
	if (!al || !bl)
		goto nomem;

	l.next = NULL;
	count = diff(al, an, bl, bn, &l);
	if (count < 0)
		goto nomem;

	/* calculate length of output */
	la = lb = 0;
	for (h = l.next; h; h = h->next) {
		if (h->a1 != la || h->b1 != lb)
			len += 12 + bl[h->b1].l - bl[lb].l;
		la = h->a2;
		lb = h->b2;
	}

	result = PyBytes_FromStringAndSize(NULL, len);

	if (!result)
		goto nomem;

	/* build binary patch */
	rb = PyBytes_AsString(result);
	la = lb = 0;

	for (h = l.next; h; h = h->next) {
		if (h->a1 != la || h->b1 != lb) {
			len = bl[h->b1].l - bl[lb].l;
			encode[0] = htonl(al[la].l - al->l);
			encode[1] = htonl(al[h->a1].l - al->l);
			encode[2] = htonl(len);
			memcpy(rb, encode, 12);
			memcpy(rb + 12, bl[lb].l, len);
			rb += 12 + len;
		}
		la = h->a2;
		lb = h->b2;
	}

nomem:
	free(al);
	free(bl);
	freehunks(l.next);
	return result ? result : PyErr_NoMemory();
}

static char mdiff_doc[] = "Efficient binary diff.";

static PyMethodDef methods[] = {
	{"bdiff", bdiff, METH_VARARGS, "calculate a binary diff\n"},
	{"blocks", blocks, METH_VARARGS, "find a list of matching lines\n"},
	{NULL, NULL}
};

#ifdef IS_PY3K
static struct PyModuleDef bdiff_module = {
	PyModuleDef_HEAD_INIT,
	"bdiff",
	mdiff_doc,
	-1,
	methods
};

PyMODINIT_FUNC PyInit_bdiff(void)
{
	return PyModule_Create(&bdiff_module);
}
#else
PyMODINIT_FUNC initbdiff(void)
{
	Py_InitModule3("bdiff", methods, mdiff_doc);
}
#endif

