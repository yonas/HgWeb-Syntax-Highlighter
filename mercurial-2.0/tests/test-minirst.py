from pprint import pprint
from mercurial import minirst

def debugformat(text, form, **kwargs):
    if form == 'html':
        print "html format:"
        out = minirst.format(text, style=form, **kwargs)
    else:
        print "%d column format:" % form
        out = minirst.format(text, width=form, **kwargs)

    print "-" * 70
    if type(out) == tuple:
        print out[0][:-1]
        print "-" * 70
        pprint(out[1])
    else:
        print out[:-1]
    print "-" * 70
    print

def debugformats(title, text, **kwargs):
    print "== %s ==" % title
    debugformat(text, 60, **kwargs)
    debugformat(text, 30, **kwargs)
    debugformat(text, 'html', **kwargs)

paragraphs = """
This is some text in the first paragraph.

  A small indented paragraph.
  It is followed by some lines
  containing random whitespace.
 \n  \n   \nThe third and final paragraph.
"""

debugformats('paragraphs', paragraphs)

definitions = """
A Term
  Definition. The indented
  lines make up the definition.
Another Term
  Another definition. The final line in the
   definition determines the indentation, so
    this will be indented with four spaces.

  A Nested/Indented Term
    Definition.
"""

debugformats('definitions', definitions)

literals = r"""
The fully minimized form is the most
convenient form::

  Hello
    literal
      world

In the partially minimized form a paragraph
simply ends with space-double-colon. ::

  ////////////////////////////////////////
  long un-wrapped line in a literal block
  \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\

::

  This literal block is started with '::',
    the so-called expanded form. The paragraph
      with '::' disappears in the final output.
"""

debugformats('literals', literals)

lists = """
- This is the first list item.

  Second paragraph in the first list item.

- List items need not be separated
  by a blank line.
- And will be rendered without
  one in any case.

We can have indented lists:

  - This is an indented list item

  - Another indented list item::

      - A literal block in the middle
            of an indented list.

      (The above is not a list item since we are in the literal block.)

::

  Literal block with no indentation (apart from
  the two spaces added to all literal blocks).

1. This is an enumerated list (first item).
2. Continuing with the second item.

(1) foo
(2) bar

1) Another
2) List

Line blocks are also a form of list:

| This is the first line.
  The line continues here.
| This is the second line.
"""

debugformats('lists', lists)

options = """
There is support for simple option lists,
but only with long options:

-X, --exclude  filter  an option with a short and long option with an argument
-I, --include          an option with both a short option and a long option
--all                  Output all.
--both                 Output both (this description is
                       quite long).
--long                 Output all day long.

--par                 This option has two paragraphs in its description.
                      This is the first.

                      This is the second.  Blank lines may be omitted between
                      options (as above) or left in (as here).


The next paragraph looks like an option list, but lacks the two-space
marker after the option. It is treated as a normal paragraph:

--foo bar baz
"""

debugformats('options', options)

fields = """
:a: First item.
:ab: Second item. Indentation and wrapping
     is handled automatically.

Next list:

:small: The larger key below triggers full indentation here.
:much too large: This key is big enough to get its own line.
"""

debugformats('fields', fields)

containers = """
Normal output.

.. container:: debug

   Initial debug output.

.. container:: verbose

   Verbose output.

   .. container:: debug

      Debug output.
"""

debugformats('containers (normal)', containers)
debugformats('containers (verbose)', containers, keep=['verbose'])
debugformats('containers (debug)', containers, keep=['debug'])
debugformats('containers (verbose debug)', containers,
            keep=['verbose', 'debug'])

roles = """Please see :hg:`add`."""
debugformats('roles', roles)


sections = """
Title
=====

Section
-------

Subsection
''''''''''

Markup: ``foo`` and :hg:`help`
------------------------------
"""
debugformats('sections', sections)


admonitions = """
.. note::
   This is a note

   - Bullet 1
   - Bullet 2

   .. warning:: This is a warning Second
      input line of warning

.. danger::
   This is danger
"""

debugformats('admonitions', admonitions)

comments = """
Some text.

.. A comment

   .. An indented comment

   Some indented text.

..

Empty comment above
"""

debugformats('comments', comments)


data = [['a', 'b', 'c'],
         ['1', '2', '3'],
         ['foo', 'bar', 'baz this list is very very very long man']]

table = minirst.maketable(data, 2, True)

print table

debugformats('table', table)
