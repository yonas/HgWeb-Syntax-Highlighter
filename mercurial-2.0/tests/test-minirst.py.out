== paragraphs ==
60 column format:
----------------------------------------------------------------------
This is some text in the first paragraph.

  A small indented paragraph. It is followed by some lines
  containing random whitespace.

The third and final paragraph.
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
This is some text in the first
paragraph.

  A small indented paragraph.
  It is followed by some lines
  containing random
  whitespace.

The third and final paragraph.
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
This is some text in the first paragraph.
</p>
<p>
A small indented paragraph.
It is followed by some lines
containing random whitespace.
</p>
<p>
The third and final paragraph.
</p>
----------------------------------------------------------------------

== definitions ==
60 column format:
----------------------------------------------------------------------
A Term
  Definition. The indented lines make up the definition.

Another Term
    Another definition. The final line in the definition
    determines the indentation, so this will be indented
    with four spaces.

  A Nested/Indented Term
    Definition.
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
A Term
  Definition. The indented
  lines make up the
  definition.

Another Term
    Another definition. The
    final line in the
    definition determines the
    indentation, so this will
    be indented with four
    spaces.

  A Nested/Indented Term
    Definition.
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<dl>
 <dt>A Term
 <dd>Definition. The indented lines make up the definition.
 <dt>Another Term
 <dd>Another definition. The final line in the definition determines the indentation, so this will be indented with four spaces.
 <dt>A Nested/Indented Term
 <dd>Definition.
</dl>
----------------------------------------------------------------------

== literals ==
60 column format:
----------------------------------------------------------------------
The fully minimized form is the most convenient form:

  Hello
    literal
      world

In the partially minimized form a paragraph simply ends with
space-double-colon.

  ////////////////////////////////////////
  long un-wrapped line in a literal block
  \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\

  This literal block is started with '::',
    the so-called expanded form. The paragraph
      with '::' disappears in the final output.
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
The fully minimized form is
the most convenient form:

  Hello
    literal
      world

In the partially minimized
form a paragraph simply ends
with space-double-colon.

  ////////////////////////////////////////
  long un-wrapped line in a literal block
  \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\

  This literal block is started with '::',
    the so-called expanded form. The paragraph
      with '::' disappears in the final output.
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
The fully minimized form is the most
convenient form:
</p>
<pre>
Hello
  literal
    world
</pre>
<p>
In the partially minimized form a paragraph
simply ends with space-double-colon.
</p>
<pre>
////////////////////////////////////////
long un-wrapped line in a literal block
\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
</pre>
<pre>
This literal block is started with '::',
  the so-called expanded form. The paragraph
    with '::' disappears in the final output.
</pre>
----------------------------------------------------------------------

== lists ==
60 column format:
----------------------------------------------------------------------
- This is the first list item.

  Second paragraph in the first list item.

- List items need not be separated by a blank line.
- And will be rendered without one in any case.

We can have indented lists:

  - This is an indented list item
  - Another indented list item:

      - A literal block in the middle
            of an indented list.

      (The above is not a list item since we are in the literal block.)

  Literal block with no indentation (apart from
  the two spaces added to all literal blocks).

1. This is an enumerated list (first item).
2. Continuing with the second item.
(1) foo
(2) bar
1) Another
2) List

Line blocks are also a form of list:

This is the first line. The line continues here.
This is the second line.
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
- This is the first list item.

  Second paragraph in the
  first list item.

- List items need not be
  separated by a blank line.
- And will be rendered without
  one in any case.

We can have indented lists:

  - This is an indented list
    item
  - Another indented list
    item:

      - A literal block in the middle
            of an indented list.

      (The above is not a list item since we are in the literal block.)

  Literal block with no indentation (apart from
  the two spaces added to all literal blocks).

1. This is an enumerated list
   (first item).
2. Continuing with the second
   item.
(1) foo
(2) bar
1) Another
2) List

Line blocks are also a form of
list:

This is the first line. The
line continues here.
This is the second line.
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<ul>
 <li> This is the first list item.
<p>
Second paragraph in the first list item.
</p>
 <li> List items need not be separated   by a blank line.
 <li> And will be rendered without   one in any case.
</ul>
<p>
We can have indented lists:
</p>
<ul>
 <li> This is an indented list item
 <li> Another indented list item:
<pre>
- A literal block in the middle
      of an indented list.
</pre>
<pre>
(The above is not a list item since we are in the literal block.)
</pre>
</ul>
<pre>
Literal block with no indentation (apart from
the two spaces added to all literal blocks).
</pre>
<ol>
 <li> This is an enumerated list (first item).
 <li> Continuing with the second item.
 <li> foo
 <li> bar
 <li> Another
 <li> List
</ol>
<p>
Line blocks are also a form of list:
</p>
<ol>
 <li> This is the first line.   The line continues here.
 <li> This is the second line.
</ol>
----------------------------------------------------------------------

== options ==
60 column format:
----------------------------------------------------------------------
There is support for simple option lists, but only with long
options:

 -X --exclude filter  an option with a short and long option
                      with an argument
 -I --include         an option with both a short option and
                      a long option
    --all             Output all.
    --both            Output both (this description is quite
                      long).
    --long            Output all day long.
    --par             This option has two paragraphs in its
                      description. This is the first.

                      This is the second.  Blank lines may
                      be omitted between options (as above)
                      or left in (as here).

The next paragraph looks like an option list, but lacks the
two-space marker after the option. It is treated as a normal
paragraph:

--foo bar baz
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
There is support for simple
option lists, but only with
long options:

 -X --exclude filter  an
                      option
                      with a
                      short
                      and
                      long
                      option
                      with an
                      argumen
                      t
 -I --include         an
                      option
                      with
                      both a
                      short
                      option
                      and a
                      long
                      option
    --all             Output
                      all.
    --both            Output
                      both
                      (this d
                      escript
                      ion is
                      quite
                      long).
    --long            Output
                      all day
                      long.
    --par             This
                      option
                      has two
                      paragra
                      phs in
                      its des
                      criptio
                      n. This
                      is the
                      first.

                      This is
                      the
                      second.
                      Blank
                      lines
                      may be
                      omitted
                      between
                      options
                      (as
                      above)
                      or left
                      in (as
                      here).

The next paragraph looks like
an option list, but lacks the
two-space marker after the
option. It is treated as a
normal paragraph:

--foo bar baz
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
There is support for simple option lists,
but only with long options:
</p>
<dl>
 <dt>-X --exclude filter
 <dd>an option with a short and long option with an argument
 <dt>-I --include
 <dd>an option with both a short option and a long option
 <dt>   --all
 <dd>Output all.
 <dt>   --both
 <dd>Output both (this description is quite long).
 <dt>   --long
 <dd>Output all day long.
 <dt>   --par
 <dd>This option has two paragraphs in its description. This is the first.
<p>
This is the second.  Blank lines may be omitted between
options (as above) or left in (as here).
</p>
</dl>
<p>
The next paragraph looks like an option list, but lacks the two-space
marker after the option. It is treated as a normal paragraph:
</p>
<p>
--foo bar baz
</p>
----------------------------------------------------------------------

== fields ==
60 column format:
----------------------------------------------------------------------
a   First item.
ab  Second item. Indentation and wrapping is handled
    automatically.

Next list:

small       The larger key below triggers full indentation
            here.
much too large
            This key is big enough to get its own line.
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
a   First item.
ab  Second item. Indentation
    and wrapping is handled
    automatically.

Next list:

small       The larger key
            below triggers
            full indentation
            here.
much too large
            This key is big
            enough to get its
            own line.
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<dl>
 <dt>a
 <dd>First item.
 <dt>ab
 <dd>Second item. Indentation and wrapping is handled automatically.
</dl>
<p>
Next list:
</p>
<dl>
 <dt>small
 <dd>The larger key below triggers full indentation here.
 <dt>much too large
 <dd>This key is big enough to get its own line.
</dl>
----------------------------------------------------------------------

== containers (normal) ==
60 column format:
----------------------------------------------------------------------
Normal output.
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Normal output.
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
Normal output.
</p>
----------------------------------------------------------------------

== containers (verbose) ==
60 column format:
----------------------------------------------------------------------
Normal output.

Verbose output.
----------------------------------------------------------------------
['debug', 'debug']
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Normal output.

Verbose output.
----------------------------------------------------------------------
['debug', 'debug']
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
Normal output.
</p>
<p>
Verbose output.
</p>
----------------------------------------------------------------------
['debug', 'debug']
----------------------------------------------------------------------

== containers (debug) ==
60 column format:
----------------------------------------------------------------------
Normal output.

Initial debug output.
----------------------------------------------------------------------
['verbose']
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Normal output.

Initial debug output.
----------------------------------------------------------------------
['verbose']
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
Normal output.
</p>
<p>
Initial debug output.
</p>
----------------------------------------------------------------------
['verbose']
----------------------------------------------------------------------

== containers (verbose debug) ==
60 column format:
----------------------------------------------------------------------
Normal output.

Initial debug output.

Verbose output.

Debug output.
----------------------------------------------------------------------
[]
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Normal output.

Initial debug output.

Verbose output.

Debug output.
----------------------------------------------------------------------
[]
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
Normal output.
</p>
<p>
Initial debug output.
</p>
<p>
Verbose output.
</p>
<p>
Debug output.
</p>
----------------------------------------------------------------------
[]
----------------------------------------------------------------------

== roles ==
60 column format:
----------------------------------------------------------------------
Please see "hg add".
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Please see "hg add".
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
Please see "hg add".
</p>
----------------------------------------------------------------------

== sections ==
60 column format:
----------------------------------------------------------------------
Title
=====

Section
-------

Subsection
''''''''''

Markup: "foo" and "hg help"
---------------------------
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Title
=====

Section
-------

Subsection
''''''''''

Markup: "foo" and "hg help"
---------------------------
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<h1>Title</h1>
<h2>Section</h2>
<h3>Subsection</h3>
<h2>Markup: "foo" and "hg help"</h2>
----------------------------------------------------------------------

== admonitions ==
60 column format:
----------------------------------------------------------------------
Note:
   This is a note

   - Bullet 1
   - Bullet 2

   Warning!
      This is a warning Second input line of warning

!Danger!
   This is danger
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Note:
   This is a note

   - Bullet 1
   - Bullet 2

   Warning!
      This is a warning Second
      input line of warning

!Danger!
   This is danger
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
<b>Note:</b> This is a note
</p>
<ul>
 <li> Bullet 1
 <li> Bullet 2
</ul>
<p>
<b>Warning!</b> This is a warning Second input line of warning
</p>
<p>
<b>!Danger!</b> This is danger
</p>
----------------------------------------------------------------------

== comments ==
60 column format:
----------------------------------------------------------------------
Some text.

   Some indented text.

Empty comment above
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
Some text.

   Some indented text.

Empty comment above
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<p>
Some text.
</p>
<p>
Some indented text.
</p>
<p>
Empty comment above
</p>
----------------------------------------------------------------------

  === === ========================================
  a   b   c                                       
  === === ========================================
  1   2   3                                       
  foo bar baz this list is very very very long man
  === === ========================================

== table ==
60 column format:
----------------------------------------------------------------------
  a   b   c
  ------------------------------------------------
  1   2   3
  foo bar baz this list is very very very long man
----------------------------------------------------------------------

30 column format:
----------------------------------------------------------------------
  a   b   c
  ------------------------------
  1   2   3
  foo bar baz this list is
          very very very long
          man
----------------------------------------------------------------------

html format:
----------------------------------------------------------------------
<table>
 <tr><th>a</th><th>b</th><th>c</th></tr>
 <tr><td>1</td><td>2</td><td>3</td></tr>
 <tr><td>foo</td><td>bar</td><td>baz this list is very very very long man</td></tr>
</table>
----------------------------------------------------------------------

